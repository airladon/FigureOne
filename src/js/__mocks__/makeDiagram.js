// import {
//   DiagramElementPrimitive,
//   DiagramElementCollection,
//   // AnimationPhase,
// } from '../Element';
import Diagram from '../diagram/Diagram';
import {
  Rect, getPoint, Point,
} from '../tools/g2';
import webgl from './WebGLInstanceMock';
import DrawContext2D from './DrawContext2DMock';
// import VertexPolygon from '../DrawingObjects/VertexObject/VertexPolygon';
import * as tools from '../tools/tools';
import { round } from '../tools/math';

tools.isTouchDevice = jest.fn();

jest.mock('../diagram/Gesture');
jest.mock('../diagram/webgl/webgl');
jest.mock('../diagram/DrawContext2D');
jest.mock('../diagram/DrawingObjects/HTMLObject/HTMLObject');

const generateRandomStringMock = jest.spyOn(tools, 'generateRandomString');
generateRandomStringMock.mockImplementation(() => '000000');

export default function makeDiagram(
  inputCanvas = new Rect(100, -300, 1000, 500),
  inputLimits = new Rect(-1, -1, 2, 2),
) {
  document.body.innerHTML =
    '<div id="c">'
    + '  <canvas class="figureone__gl" id="id_figureone__gl__low">'
    + '  </canvas>'
    + '  <canvas class="figureone__text" id="id_figureone__text__low">'
    + '  </canvas>'
    + '  <div class="figureone__html">'
    + '  </div>'
    + '  <canvas class="figureone__gl" id="id_figureone__gl__high">'
    + '  </canvas>'
    + '  <canvas class="figureone__text" id="id_figureone__text__high">'
    + '  </canvas>'
    + '</div>';
  // canvas = document.getElementById('c');
  const definition = {
    width: inputCanvas.width,
    height: inputCanvas.height,
    limits: inputLimits,
  };

  const canvasMock = {
    width: definition.width,
    height: definition.height,
    // offsetLeft: 100,
    left: inputCanvas.left,
    // offsetTop: 200,
    top: inputCanvas.top,
    // width: definition.width,
    // height: definition.height,
    offsetWidth: definition.width,
    offsetHeight: definition.height,
    scrollLeft: 0,
    scrollTop: 0,
    // eslint-disable-next-line arrow-body-style
    getBoundingClientRect: () => {
      return {
        left: inputCanvas.left,
        top: inputCanvas.top,
        width: definition.width,
        height: definition.height,
      };
    },
  };
  const htmlCanvasMock = {
    style: {
      fontsize: 1,
    },
    offsetWidth: 100,
    appendChild: () => {},
  };
  const { limits } = definition;
  const diagram = new Diagram({ htmlId: 'c', limits });
  diagram.webglLow = webgl;
  diagram.webglHigh = webgl;
  diagram.webgl = webgl;
  diagram.canvasLow = canvasMock;
  diagram.canvasHigh = canvasMock;
  diagram.htmlCanvas = htmlCanvasMock;
  diagram.isTouchDevice = false;
  diagram.draw2DLow = new DrawContext2D(definition.width, definition.height);
  diagram.draw2DHigh = new DrawContext2D(definition.width, definition.height);
  diagram.draw2D = diagram.draw2DLow;
  diagram.shapesLow = diagram.getShapes(false);
  diagram.shapesHigh = diagram.getShapes(true);
  diagram.shapes = diagram.shapesLow;
  diagram.equationLow = diagram.getEquations(false);
  diagram.equationHigh = diagram.getEquations(true);
  diagram.equation = diagram.equationLow;
  diagram.advancedLow = diagram.getObjects(false);
  diagram.advancedHigh = diagram.getObjects(true);
  diagram.advanced = diagram.advancedLow;
  diagram.setSpaceTransforms();
  // needed as the first element needs to be set with the space Transforms
  diagram.initElements();
  diagram.mock = {
    initialTime: 0,
    duration: 0,
    previousTouchPoint: new Point(0, 0),
    timersBeforeDraw: true,
    timeStep: (deltaTimeInSecondsIn, frameTimeIn = null) => {
      const { duration, initialTime } = diagram.mock;
      const deltaTimeInSeconds = round(deltaTimeInSecondsIn, 8);
      let frameTime = deltaTimeInSeconds;
      // let deltaTime = 0;
      if (frameTimeIn != null && frameTimeIn < deltaTimeInSeconds) {
        frameTime = frameTimeIn;
        // deltaTime = frameTime;
      }
      let deltaTime = frameTime;

      // let delta = Math.min(deltaTimeInSeconds, frameTime);
      let lastTime = 0;
      deltaTime = round(deltaTime, 8);
      while (deltaTime <= round(deltaTimeInSeconds, 8)) {
        // if (window.asdf) {
        //   console.log(duration, deltaTime + 0.00);
        // }
        const newNow = round((duration + deltaTime + initialTime) * 1000, 8);
        global.performance.now = () => newNow;
        if (diagram.mock.timersBeforeDraw) {
          jest.advanceTimersByTime(round((deltaTime - lastTime) * 1000, 8));
        }
        diagram.animateNextFrame();
        diagram.draw(round(newNow / 1000, 8));
        if (!diagram.mock.timersBeforeDraw) {
          jest.advanceTimersByTime(round((deltaTime - lastTime) * 1000, 8));
        }
        lastTime = deltaTime;
        if (deltaTime === deltaTimeInSeconds || frameTime === 0) {
          deltaTime += 1;
        } else if (deltaTime + frameTime > deltaTimeInSeconds) {
          deltaTime = deltaTimeInSeconds;
        } else {
          deltaTime += frameTime;
        }
        deltaTime = round(deltaTime, 8);
      }
      diagram.mock.duration += deltaTimeInSeconds;
    },
    touchDown: (diagramPosition) => {
      const p = getPoint(diagramPosition);
      const pixelPoint = p.transformBy(diagram.spaceTransforms.diagramToPixel.m());
      const clientPoint = diagram.pixelToClient(pixelPoint);
      diagram.touchDownHandler(clientPoint);
      diagram.mock.previousTouchPoint = clientPoint;
    },
    touchUp: () => {
      diagram.touchUpHandler();
    },
    touchMove: (diagramPosition) => {
      const p = getPoint(diagramPosition);
      const pixelPoint = p.transformBy(diagram.spaceTransforms.diagramToPixel.m());
      const clientPoint = diagram.pixelToClient(pixelPoint);
      diagram.touchMoveHandler(diagram.mock.previousTouchPoint, clientPoint);
      diagram.mock.previousTouchPoint = clientPoint;
    },
  };
  diagram.globalAnimation.reset();
  return diagram;
}

