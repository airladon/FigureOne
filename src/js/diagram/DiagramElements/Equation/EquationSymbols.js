// @flow
import {
  Point, Transform,
} from '../../../tools/g2';
import { joinObjects } from '../../../tools/tools';
import DiagramPrimitives from '../../DiagramPrimitives/DiagramPrimitives';
// import { DiagramElementCollection } from '../../Element';
import Integral from './Symbols/Integral';
// import SuperSub from './Elements/SuperSub';
import Bracket from './Symbols/Bracket';
import Box from './Symbols/Box';
import Radical from './Symbols/Radical';
import Brace from './Symbols/Brace';
import SquareBracket from './Symbols/SquareBracket';
import AngleBracket from './Symbols/AngleBracket';
// import SquareBracketNew from './Symbols/SquareBracketNew';
import Bar from './Symbols/Bar';
import Sum from './Symbols/Sum';
import Product from './Symbols/Product';
import SimpleIntegral from './Symbols/SimpleIntegral';
import IntegralNew from './Symbols/IntegralNew';
// import BracketNew from './Symbols/BracketNew';
// import BraceNew from './Symbols/BraceNew';

// import { Annotation, AnnotationInformation } from './Elements/Annotation';
export type TypeSymbolOptions = {
  color?: Array<number>,
  numLines?: number,
  side?: 'left' | 'right' | 'bottom' | 'top',
  width?: number,
  fill?: boolean,
  staticSize?: ?(Point | [number, number] | number),
  startWidth?: number,
  lineWidth?: number,
  startHeight?: number,
  maxStartWidth?: ?number,
  maxStartHeight?: ?number,
  proportionalToHeight?: boolean,
  endLength?: number,
  sides?: number,
  tipWidth?: number,
  radius?: number,
}

export default class EquationSymbols {
  shapes: DiagramPrimitives;
  defaultColor: Array<number>;

  constructor(
    shapes: DiagramPrimitives,
    defaultColor: Array<number>,
  ) {
    this.shapes = shapes;
    this.defaultColor = defaultColor;
  }

  get(
    name: string,
    options: TypeSymbolOptions,
  ) {
    if (name === 'vinculum') {
      return this.vinculum(options);
    }
    if (name === 'strike') {
      return this.strike(options);
    }
    if (name === 'xStrike') {
      return this.xStrike(options);
    }
    if (name === 'integral') {
      return this.integral(options);
    }
    if (name === 'bracket') {         // $FlowFixMe
      return this.bracket(options);
    }
    if (name === 'squareBracket') {         // $FlowFixMe
      return this.squareBracket(options);
    }
    if (name === 'brace') {         // $FlowFixMe
      return this.brace(options);
    }
    if (name === 'bar') {         // $FlowFixMe
      return this.bar(options);
    }
    if (name === 'box') {         // $FlowFixMe
      return this.box(options);
    }
    if (name === 'angleBracket') {         // $FlowFixMe
      return this.angleBracket(options);
    }
    if (name === 'radical') {         // $FlowFixMe
      return this.radical(options);
    }
    if (name === 'simpleIntegral') {
      return this.simpleIntegral(options);
    }
    if (name === 'sum') {
      return this.sum(options);
    }
    if (name === 'prod') {
      return this.product(options);
    }
    if (name === 'int') {
      return this.integralNew(options);
    }
    return null;
  }

  vinculum(options: { color?: Array<number> } = {}) {
    let { color } = options;
    if (color == null) {
      color = this.defaultColor;
    }
    return this.shapes.horizontalLine(
      new Point(0, 0),
      1, 1, 0,
      color,
      new Transform('vinculum').scale(1, 1).translate(0, 0),
    );
  }

  box(optionsIn: {
    color?: Array<number>,
    fill?: boolean,
    width?: number,
    staticSize?: ?(Point | [number, number]),
  }) {
    const defaultOptions = {
      color: this.defaultColor,
      fill: false,
      width: 0.01,
      staticSize: null,
    };
    const options = joinObjects(defaultOptions, optionsIn);
    return Box(this.shapes, options.color, options.fill, options.width, options.staticSize);
  }

  // simpleIntegral(optionsIn: {
  //   color?: Array<number>,
  //   lineWidth?: number,
  //   staticSize?: ?(Point | [number, number]),
  // }) {
  //   const defaultOptions: {
  //     color: this.defaultColor,
  //     lineWidth: 0.01,
  //   }
  // }

  simpleIntegral(options: {
    color?: Array<number>,
    lineWidth?: number,
    staticSize?: boolean,
  }) {
    const defaultOptions = {
      color: this.defaultColor,
      lineWidth: 0.012,
      staticSize: null,
    };
    const optionsToUse = joinObjects(defaultOptions, options);

    return (new SimpleIntegral(
      this.shapes.webgl,
      optionsToUse.color,
      new Transform('brace').scale(1, 1).translate(0, 0),
      this.shapes.limits,
      optionsToUse.staticSize,
      {
        lineWidth: optionsToUse.lineWidth,
      },
    )).symbol;
  }

  sum(options: {
    color?: Array<number>,
    lineWidth?: number,
    sides?: number,
    staticSize?: boolean,
  }) {
    const defaultOptions = {
      color: this.defaultColor,
      lineWidth: null,
      sides: 5,
      staticSize: true,
    };
    const optionsToUse = joinObjects(defaultOptions, options);

    return (new Sum(
      this.shapes.webgl,
      optionsToUse.color,
      new Transform('brace').scale(1, 1).translate(0, 0),
      this.shapes.limits,
      optionsToUse.staticSize,
      {
        lineWidth: optionsToUse.lineWidth,
        sides: optionsToUse.sides,
      },
    )).symbol;
  }

  product(options: {
    color?: Array<number>,
    lineWidth?: number,
    sides?: number,
    staticSize?: boolean,
  }) {
    const defaultOptions = {
      color: this.defaultColor,
      lineWidth: null,
      sides: 5,
      staticSize: true,
    };
    const optionsToUse = joinObjects(defaultOptions, options);

    return (new Product(
      this.shapes.webgl,
      optionsToUse.color,
      new Transform('brace').scale(1, 1).translate(0, 0),
      this.shapes.limits,
      optionsToUse.staticSize,
      {
        lineWidth: optionsToUse.lineWidth,
        sides: optionsToUse.sides,
      },
    )).symbol;
  }

  integralNew(options: {
    color?: Array<number>,
    lineWidth?: number,
    sides?: number,
    width?: ?number,
    tipWidth?: number,
    percentage?: number,
    staticSize?: boolean,
    serif?: boolean,
  }) {
    const defaultOptions = {
      color: this.defaultColor,
      lineWidth: null,
      width: null,
      tipWidth: null,
      percentage: 0.95,
      sides: 5,
      staticSize: true,
      radius: 0.03,
      serif: true,
    };
    const optionsToUse = joinObjects(defaultOptions, options);

    return (new IntegralNew(
      this.shapes.webgl,
      optionsToUse.color,
      new Transform('brace').scale(1, 1).translate(0, 0),
      this.shapes.limits,
      optionsToUse.staticSize,
      optionsToUse,
      // {
      //   lineWidth: optionsToUse.lineWidth,
      //   minLineWidth: optionsToUse.minLineWidth,
      //   sides: optionsToUse.sides,
      //   width: optionsToUse.width,
      //   percentage: optionsToUse.percentage,
      //   sigma: optionsToUse.sigma,
      // },
    )).symbol;
  }

  radical(optionsIn: {
    color?: Array<number>,
    lineWidth?: number,
    startHeight?: number,
    startWidth?: number,
    proportionalToHeight?: boolean,
    maxStartWidth?: ?number,
    maxStartHeight?: ?number,
    staticSize?: ?(Point | [number, number]),
  }) {
    const defaultOptions: {
      color: Array<number>,
      lineWidth: number,
      startHeight: number,
      startWidth: number,
      proportionalToHeight: boolean,
      maxStartWidth: ?number,
      maxStartHeight: ?number,
      staticSize: ?(Point | [number, number]),
    } = {
      color: this.defaultColor,
      lineWidth: 0.01,
      staticSize: null,
      startHeight: 0.5,
      startWidth: 0.7,
      maxStartWidth: 0.15,
      maxStartHeight: 0.15,
      proportionalToHeight: true,
    };
    if (optionsIn.proportionalToHeight != null
      && optionsIn.proportionalToHeight === false
    ) {
      defaultOptions.startHeight = 0.15;
      defaultOptions.startWidth = 0.15;
      defaultOptions.maxStartHeight = null;
      defaultOptions.maxStartWidth = null;
    }
    const options = joinObjects(defaultOptions, optionsIn);
    return Radical(
      this.shapes, options.color, options.lineWidth,
      options.startWidth, options.startHeight, options.proportionalToHeight,
      options.maxStartWidth, options.maxStartHeight,
      options.staticSize,
    );
  }

  strike(options: { color?: Array<number> } = {}) {
    let { color } = options;
    if (color == null) {
      color = this.defaultColor;
    }
    return this.shapes.horizontalLine(
      new Point(0, 0),
      1, 1, 0,
      color,
      new Transform('strike').scale(1, 1).rotate(0).translate(0, 0),
    );
  }

  xStrike(options: { color?: Array<number> } = {}) {
    let { color } = options;
    if (color == null) {
      color = this.defaultColor;
    }
    const cross = this.shapes.collection(new Transform('xStrike').scale(1, 1).rotate(0).translate(0, 0));
    cross.color = color;
    const strike1 = this.shapes.horizontalLine(
      new Point(0, 0),
      1, 1, 0,
      color,
      new Transform('strikeLine').scale(1, 1).rotate(0).translate(0, 0),
    );
    const strike2 = strike1._dup();
    cross.add('s1', strike1);
    cross.add('s2', strike2);
    return cross;
  }

  integral(options: { color?: Array<number>, numLines?: number}) {
    const defaultOptions = {
      color: this.defaultColor,
      numLines: 1,
    };
    const optionsToUse = joinObjects(defaultOptions, options);
    return new Integral(
      this.shapes.webgl,
      optionsToUse.color,
      optionsToUse.numLines,
      new Transform('integral').scale(1, 1).translate(0, 0),
      this.shapes.limits,
    );
  }

  bracket(options: {
    side?: 'left' | 'right' | 'top' | 'bottom',
    color?: Array<number>,
    lineWidth?: number,
    sides?: number,
    width?: number,
    tipWidth?: number,
    staticSize?: boolean,
  }) {
    const defaultOptions = {
      side: 'left',
      color: this.defaultColor,
      lineWidth: 0.012,
      sides: 10,
      staticSize: null,
    };
    const optionsToUse = joinObjects(defaultOptions, options);
    if (optionsToUse.width == null) {
      optionsToUse.width = optionsToUse.lineWidth * 2.5;
    }
    if (optionsToUse.tipWidth == null) {
      optionsToUse.tipWidth = optionsToUse.lineWidth / 3;
    }
    return (new Bracket(
      this.shapes.webgl,
      optionsToUse.color,
      new Transform('bracket').scale(1, 1).translate(0, 0),
      this.shapes.limits,
      optionsToUse.side,
      optionsToUse.staticSize,
      {
        sides: optionsToUse.sides,
        lineWidth: optionsToUse.lineWidth,
        width: optionsToUse.width,
        tipWidth: optionsToUse.tipWidth,
      },
    )).symbol;
  }

  brace(options: {
    side?: 'left' | 'right' | 'top' | 'bottom',
    color?: Array<number>,
    lineWidth?: number,
    sides?: number,
    width?: number,
    tipWidth?: number,
    staticSize?: boolean,
  }) {
    const defaultOptions = {
      side: 'left',
      color: this.defaultColor,
      lineWidth: 0.012,
      sides: 10,
      staticSize: null,
    };
    const optionsToUse = joinObjects(defaultOptions, options);
    if (optionsToUse.width == null) {
      optionsToUse.width = optionsToUse.lineWidth * 4.2;
    }
    if (optionsToUse.tipWidth == null) {
      optionsToUse.tipWidth = optionsToUse.lineWidth / 3;
    }
    return (new Brace(
      this.shapes.webgl,
      optionsToUse.color,
      new Transform('brace').scale(1, 1).translate(0, 0),
      this.shapes.limits,
      optionsToUse.side,
      optionsToUse.staticSize,
      {
        sides: optionsToUse.sides,
        lineWidth: optionsToUse.lineWidth,
        width: optionsToUse.width,
        tipWidth: optionsToUse.tipWidth,
      },
    )).symbol;
  }

  bar(options: {
    side?: 'left' | 'right' | 'top' | 'bottom',
    color?: Array<number>,
    lineWidth?: number,
    staticSize?: ?boolean,
  }) {
    const defaultOptions = {
      side: 'left',
      lineWidth: 0.01,
      color: this.defaultColor,
      staticSize: null,
    };
    const optionsToUse = joinObjects(defaultOptions, options);
    return (new Bar(
      this.shapes.webgl,
      optionsToUse.color,
      new Transform('bar').scale(1, 1).translate(0, 0),
      this.shapes.limits,
      optionsToUse.side,
      optionsToUse.staticSize,
      {
        lineWidth: optionsToUse.lineWidth,
      },
    )).symbol;
  }

  squareBracket(options: {
    side?: 'left' | 'right' | 'top' | 'bottom',
    color?: Array<number>,
    lineWidth?: number,
    width?: number,
    endLineWidth?: number,
    staticSize?: ?boolean,
    radius?: number,
    sides?: number,
  }) {
    const defaultOptions = {
      side: 'left',
      lineWidth: 0.01,
      color: this.defaultColor,
      width: 0.04,
      staticSize: null,
      radius: 0,
      sides: 5,
    };
    const optionsToUse = joinObjects(defaultOptions, options);
    if (optionsToUse.endLineWidth == null) {
      optionsToUse.endLineWidth = optionsToUse.lineWidth * 0.7;
    }
    return (new SquareBracket(
      this.shapes.webgl,
      optionsToUse.color,
      new Transform('bar').scale(1, 1).translate(0, 0),
      this.shapes.limits,
      optionsToUse.side,
      optionsToUse.staticSize,
      {
        lineWidth: optionsToUse.lineWidth,
        width: optionsToUse.width,
        endLineWidth: optionsToUse.endLineWidth,
        radius: optionsToUse.radius,
        sides: optionsToUse.sides,
      },
    )).symbol;
  }

  angleBracket(options: {
    side?: 'left' | 'right' | 'top' | 'bottom',
    color?: Array<number>,
    lineWidth?: number,
    width?: number,
    staticSize?: ?boolean,
  }) {
    const defaultOptions = {
      side: 'left',
      lineWidth: 0.01,
      width: 0.05,
      color: this.defaultColor,
      staticSize: null,
    };
    const optionsToUse = joinObjects(defaultOptions, options);
    return (new AngleBracket(
      this.shapes.webgl,
      optionsToUse.color,
      new Transform('bar').scale(1, 1).translate(0, 0),
      this.shapes.limits,
      optionsToUse.side,
      optionsToUse.staticSize,
      {
        lineWidth: optionsToUse.lineWidth,
        width: optionsToUse.width,
      },
    )).symbol;
  }
}
