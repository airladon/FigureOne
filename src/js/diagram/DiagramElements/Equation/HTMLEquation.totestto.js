// import HTMLEquation from './HTMLEquation';

// describe('HTML Equation', () => {
//   test('Text', () => {
//     const t = new e.Text('test', 'id1', 'c1');
//     const rendered = t.render();
//     const expected = '<div id="id1" class="equation_element c1 equation_text">\n  test\n</div>';
//     expect(rendered).toBe(expected);
//   });
//   test('Line', () => {
//     const t1 = new e.Text('a+b', 'id1', 'c1');
//     const t2 = new e.Text('=', 'id2', 'c2');
//     const t3 = new e.Text('c', 'id3', 'c3');

//     const expected =
// `<div id="id_e1" class="equation_element class_e1">
//   <div id="id1" class="equation_element c1 equation_text">
//     a+b
//   </div>
//   <div id="id2" class="equation_element c2 equation_text">
//     =
//   </div>
//   <div id="id3" class="equation_element c3 equation_text">
//     c
//   </div>
// </div>`;
//     const line = new e.Line(
//       [t1, t2, t3],
//       'id_e1',
//       'class_e1',
//     );
//     expect(line.render()).toBe(expected);
//   });

//   test('Fraction', () => {
//     const n = new e.Line(
//       [new e.Text('a + b')],
//       'id_n1',
//       'class_n1',
//     );
//     const d = new e.Line(
//       [new e.Text('c')],
//       'id_d1',
//       'class_d1',
//     );
//     const f = new e.Fraction(
//       n,
//       d,
//       'id_f1',
//       'class_f1',
//     );
//     const expected =
// `<div id="id_f1" class="equation_element class_f1 fraction">
//   <div class="numerator">
//     <div id="id_n1" class="equation_element class_n1">
//       <div class="equation_element equation_text">
//         a + b
//       </div>
//     </div>
//   </div>
//   <div class="fraction_line"> </div>
//   <div class="denominator">
//     <div id="id_d1" class="equation_element class_d1">
//       <div class="equation_element equation_text">
//         c
//       </div>
//     </div>
//   </div>
// </div>`;
//     expect(f.render()).toBe(expected);
//   });
//   describe('Equation', () => {
//     test('a + b = c', () => {
//       // const line = new e.Line();
//       const eq = new e.Equation('e1', [new e.Text('a + b = c')]);
//       const expected =
// `<div id="e1" class="equation_element equation">
//   <div class="equation_element equation_text">
//     a + b = c
//   </div>
// </div>`;
//       expect(eq.render()).toBe(expected);
//     });
//   });

//   test.only('temp', () => {
//     // console.log(e.e([e.e('b'), e.e('&#177;'), e.sqrt('4ac')]).render());
//     const n = e.e('b');
//     const d = e.e('2');
//     const eq = new e.Equation([e.frac(n, d), e.e('='), e.e('0')]);
//     console.log(eq.render());
//     expect(true).toBe(true);
//   })
// });

// // b +- sqrt(4ac)
// // -------------- = d
// //     2A

// // E = e.e([
// //   e.frac([e.e('b') e.e('+') e.sqrt('4ac')], e.e('2A')),
// //   e.('='),
// //   e.('d')])

// //  / inf
// //  |
// //  | x^2 dx = 2x
// //  |
// // / 0


// // E = e([e.o('&int;').ss('0', 'inf') e.t('x').sp('2') e.t('dx') e.('=') e.('2x')])
