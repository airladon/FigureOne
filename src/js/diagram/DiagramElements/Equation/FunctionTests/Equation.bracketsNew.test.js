// import {
//   Point,
// } from '../../../../tools/g2';
import {
  round,
} from '../../../../tools/math';
import * as tools from '../../../../tools/tools';
import makeDiagram from '../../../../__mocks__/makeDiagram';
import { EquationNew } from '../Equation';

tools.isTouchDevice = jest.fn();

jest.mock('../../../Gesture');
jest.mock('../../../webgl/webgl');
jest.mock('../../../DrawContext2D');

describe('Equation Functions - Brackets', () => {
  let diagram;
  let eqn;
  let color1;
  let elements;
  let functions;
  // let forms;
  beforeEach(() => {
    diagram = makeDiagram();
    color1 = [0.95, 0, 0, 1];
    elements = {
      a: 'a',
      b: 'b',
      lb: {
        symbol: 'squareBracketNew', side: 'left', lineWidth: 0.01, endLength: 0.03,
      },
      rb: {
        symbol: 'squareBracketNew', side: 'right', lineWidth: 0.01, endLength: 0.03,
      },
    };
    functions = {
      single: () => {
        eqn = new EquationNew(diagram.shapes, { color: color1 });
        const e = eqn.eqn.functions;
        const bracNew = e.bracNew.bind(e);
        eqn.addElements(elements);
        eqn.addForms({
          // Full Object
          0: {
            content: {
              bracNew: {
                content: 'a',
                left: 'lb',
                right: 'rb',
                insideSpace: 0.1,
                outsideSpace: 0.1,
                topSpace: 0.1,
                bottomSpace: 0.1,
              },
            },
          },
          // Method Object
          1: {
            bracNew: {
              content: 'a',
              left: 'lb',
              right: 'rb',
              insideSpace: 0.1,
              outsideSpace: 0.1,
              topSpace: 0.1,
              bottomSpace: 0.1,
            },
          },
          // Method Array
          2: { bracNew: ['a', 'lb', 'rb', 0.1, 0.1, 0.1, 0.1] },
          // Function with Method Array
          3: e.bracNew(['a', 'lb', 'rb', 0.1, 0.1, 0.1, 0.1]),
          // Function with parameters
          4: e.bracNew('a', 'lb', 'rb', 0.1, 0.1, 0.1, 0.1),
          // Bound Function with parameters
          5: bracNew('a', 'lb', 'rb', 0.1, 0.1, 0.1, 0.1),
          // Bound Function with Object
          6: bracNew({
            content: 'a',
            left: 'lb',
            right: 'rb',
            insideSpace: 0.1,
            outsideSpace: 0.1,
            topSpace: 0.1,
            bottomSpace: 0.1,
          }),
        });
      },
      parameterSteps: () => {
        eqn = new EquationNew(diagram.shapes, { color: color1 });
        const e = eqn.eqn.functions;
        const bracNew = e.bracNew.bind(e);
        eqn.addElements(elements);
        eqn.addForms({
          // without
          //   // Method Object
          base: {
            content: [
              {
                bracNew: {
                  content: 'a',
                  left: 'lb',
                  right: 'rb',
                  insideSpace: 0.1,
                  outsideSpace: 0.1,
                  topSpace: 0.1,
                  bottomSpace: 0.1,
                  minContentHeight: null,
                  minContentDescent: null,
                  height: null,
                  descent: null,
                  inSize: null,
                },
              },
              'b',
            ],
            scale: 1,
          },
          insideSpace: {
            content: [bracNew(
              'a', 'lb', 'rb', 0.2, 0.1, 0.1, 0.1, null, null, null, null, null,
            ), 'b'],
            scale: 1,
          },
          outsideSpace: {
            content: [bracNew(
              'a', 'lb', 'rb', 0.1, 0.2, 0.1, 0.1, null, null, null, null, null,
            ), 'b'],
            scale: 1,
          },
          topSpace: {
            content: [bracNew(
              'a', 'lb', 'rb', 0.1, 0.1, 0.2, 0.1, null, null, null, null, null,
            ), 'b'],
            scale: 1,
          },
          bottomSpace: {
            content: [bracNew(
              'a', 'lb', 'rb', 0.1, 0.1, 0.1, 0.2, null, null, null, null, null,
            ), 'b'],
            scale: 1,
          },
        });
      },
    };
  });
  describe('Parameter Steps', () => {
    let baseB;
    let baseL;
    let baseR;
    let baseA;
    let baseLScale;
    beforeEach(() => {
      functions.parameterSteps();
      eqn.showForm('base');
      baseA = eqn._a.getPosition();
      baseB = eqn._b.getPosition();
      baseL = eqn._lb.getPosition();
      baseR = eqn._rb.getPosition();
      baseLScale = eqn._lb.custom.scale._dup();
    });
    test('bottomSpace', () => {
      eqn.showForm('bottomSpace');
      const newLScale = eqn._lb.custom.scale._dup();
      const newL = eqn._lb.getPosition();
      expect(round(baseLScale.y + 0.1)).toBe(round(newLScale.y));
      expect(round(baseL.y - 0.1)).toBe(round(newL.y));
    });
    test('topSpace', () => {
      eqn.showForm('topSpace');
      const newLScale = eqn._lb.custom.scale._dup();
      const newL = eqn._lb.getPosition();
      expect(round(baseLScale.y + 0.1)).toBe(round(newLScale.y));
      expect(round(baseL.y)).toBe(round(newL.y));
    });
    test('outsideSpace', () => {
      eqn.showForm('outsideSpace');
      const newB = eqn._b.getPosition();
      const newL = eqn._lb.getPosition();
      const newR = eqn._rb.getPosition();
      const newA = eqn._a.getPosition();
      expect(round(baseL.x + 0.1)).toBe(round(newL.x));
      expect(round(baseA.x + 0.1)).toBe(round(newA.x));
      expect(round(baseR.x + 0.1)).toBe(round(newR.x));
      expect(round(baseB.x + 0.2)).toBe(round(newB.x));
    });
    test('insideSpace', () => {
      eqn.showForm('insideSpace');
      const newB = eqn._b.getPosition();
      const newL = eqn._lb.getPosition();
      const newR = eqn._rb.getPosition();
      const newA = eqn._a.getPosition();
      expect(round(baseL.x)).toBe(round(newL.x));
      expect(round(baseA.x + 0.1)).toBe(round(newA.x));
      expect(round(baseR.x + 0.2)).toBe(round(newR.x));
      expect(round(baseB.x + 0.2)).toBe(round(newB.x));
    });
  });
  test('Bracket', () => {
    functions.single();
    const elems = [eqn._a, eqn._lb, eqn._rb];
    const formsToTest = ['1', '2', '3', '4', '5', '6'];
    eqn.showForm('0');
    const positions0 = elems.map(elem => round(elem.transform.mat).slice());
    formsToTest.forEach((f) => {
      eqn.showForm(f);
      const positions = elems.map(elem => round(elem.transform.mat).slice());
      expect(positions0).toEqual(positions);
    });

    // Snapshot test on most simple layout
    eqn.showForm('0');
    tools.cleanUIDs(eqn);
    expect(round(eqn._a.transform.mat)).toMatchSnapshot();
    expect(round(eqn._lb.transform.mat)).toMatchSnapshot();
    expect(round(eqn._rb.transform.mat)).toMatchSnapshot();
  });
});
