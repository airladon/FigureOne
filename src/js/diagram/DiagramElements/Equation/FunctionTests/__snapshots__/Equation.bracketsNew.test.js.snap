// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Equation Functions - Brackets Bracket 1`] = `
Array [
  0.7,
  0,
  0.182,
  0,
  0.7,
  0,
  0,
  0,
  1,
]
`;

exports[`Equation Functions - Brackets Bracket 2`] = `
Array [
  0.2121,
  0,
  0.07,
  0,
  0.2121,
  -0.0756,
  0,
  0,
  1,
]
`;

exports[`Equation Functions - Brackets Bracket 3`] = `
Array [
  0.2121,
  0,
  0.28,
  0,
  0.2121,
  -0.0756,
  0,
  0,
  1,
]
`;
